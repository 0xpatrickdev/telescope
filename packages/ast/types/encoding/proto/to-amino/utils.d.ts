import * as t from '@babel/types';
import { ProtoType } from '@cosmology/types';
import { ProtoParseContext } from '../../context';
import { ToAminoJSONMethod } from './index';
export declare const toAminoJSON: {
    scalar(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    omitDefaultExpr(args: ToAminoJSONMethod, expr: any): t.CallExpression;
    string(args: ToAminoJSONMethod, omitEmpty: any): t.ExpressionStatement;
    double(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    float(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    bool(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    number(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    int32(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    uint32(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    sint32(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    fixed32(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    sfixed32(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    long(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    int64(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    uint64(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    sint64(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    fixed64(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    sfixed64(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    protoType(args: ToAminoJSONMethod): t.ExpressionStatement;
    anyType(args: ToAminoJSONMethod): t.ExpressionStatement;
    type(args: ToAminoJSONMethod): t.ExpressionStatement;
    enum(args: ToAminoJSONMethod, omitEmpty: boolean): t.ExpressionStatement;
    bytes(args: ToAminoJSONMethod): t.ExpressionStatement;
    duration(args: ToAminoJSONMethod): t.ExpressionStatement;
    timestamp(args: ToAminoJSONMethod): t.ExpressionStatement;
    pubkey(args: ToAminoJSONMethod): t.ExpressionStatement;
    rawBytes(args: ToAminoJSONMethod): t.ExpressionStatement;
    wasmByteCode(args: ToAminoJSONMethod): t.ExpressionStatement;
    keyHash(args: ToAminoJSONMethod): (t.ExpressionStatement | t.IfStatement)[];
    array(args: ToAminoJSONMethod, expr: t.Expression): t.IfStatement;
};
export declare const arrayTypes: {
    scalar(): t.Identifier;
    string(): t.Identifier;
    double(): t.Identifier;
    float(): t.Identifier;
    bool(): t.Identifier;
    number(): t.Identifier;
    int32(): t.Identifier;
    uint32(): t.Identifier;
    sint32(): t.Identifier;
    fixed32(): t.Identifier;
    sfixed32(): t.Identifier;
    long(args: ToAminoJSONMethod): t.Expression;
    int64(args: ToAminoJSONMethod): t.Expression;
    uint64(args: ToAminoJSONMethod): t.Expression;
    sint64(args: ToAminoJSONMethod): t.Expression;
    fixed64(args: ToAminoJSONMethod): t.Expression;
    sfixed64(args: ToAminoJSONMethod): t.Expression;
    bytes(args: ToAminoJSONMethod): t.Identifier;
    enum(args: ToAminoJSONMethod): t.CallExpression;
    anyType(args: ToAminoJSONMethod): t.ConditionalExpression;
    protoType(args: ToAminoJSONMethod): t.ConditionalExpression;
    type(args: ToAminoJSONMethod): t.ConditionalExpression;
};
export declare const toAminoMessages: {
    anyType(): (t.VariableDeclaration | t.ExpressionStatement | t.ReturnStatement)[];
    timestamp(context: ProtoParseContext, name: string, proto: ProtoType): t.ReturnStatement;
    duration(context: ProtoParseContext, name: string, proto: ProtoType): t.ReturnStatement;
};
