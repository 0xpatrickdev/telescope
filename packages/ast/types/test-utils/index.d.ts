import { ProtoType } from '@osmonauts/types';
import { ProtoStore } from '@osmonauts/proto-parser';
import { AminoParseContext } from '../src/encoding/context';
export declare const expectCode: (ast: any) => void;
export declare const printCode: (ast: any) => void;
export declare const prepareContext: (store: ProtoStore, protoFile: string) => {
    context: AminoParseContext;
    root: import("@osmonauts/types").ProtoRoot;
    protos: ProtoType[];
};
