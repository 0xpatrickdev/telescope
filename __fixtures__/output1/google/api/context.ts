import { _m0 } from "protobuf/minimal";
import { isSet } from "@osmonauts/helpers";
export interface Context {
  rules: ContextRule[];
}

function createBaseContext(): Context {
  return {
    rules: []
  };
}

export const Context = {
  encode(message: Context, writer = _m0.Writer.create()): _m0.Writer {
    for (const v of message.rules) {
      ContextRule.encode(v!, writer.uint32(10).fork()).ldelim();
    }

    return writer;
  },

  decode(input: _m0.Reader | Uint8Array): Context {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseContext();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.rules.push(ContextRule.decode(reader, reader.uint32()));
          break;

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object: any): Context {
    return {
      rules: Array.isArray(object?.rules) ? object.rules.map((e: any) => ContextRule.fromJSON(e)) : []
    };
  },

  toJSON(message: Context): unknown {
    const obj: any = {};

    if (message.rules) {
      obj.rules = message.rules.map(e => e ? ContextRule.toJSON(e) : undefined);
    } else {
      obj.rules = [];
    }

    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<Context>, I>>(object: I): Context {
    const message = createBaseContext();
    message.rules = object.rules?.map(e => ContextRule.fromPartial(e)) || [];
    return message;
  }

};
export interface ContextRule {
  selector: string;
  requested: string[];
  provided: string[];
  allowedRequestExtensions: string[];
  allowedResponseExtensions: string[];
}

function createBaseContextRule(): ContextRule {
  return {
    selector: "",
    requested: [],
    provided: [],
    allowedRequestExtensions: [],
    allowedResponseExtensions: []
  };
}

export const ContextRule = {
  encode(message: ContextRule, writer = _m0.Writer.create()): _m0.Writer {
    if (message.selector !== "") {
      writer.uint32(10).string(message.selector);
    }

    for (const v of message.requested) {
      writer.uint32(18).string(v!);
    }

    for (const v of message.provided) {
      writer.uint32(26).string(v!);
    }

    for (const v of message.allowedRequestExtensions) {
      writer.uint32(34).string(v!);
    }

    for (const v of message.allowedResponseExtensions) {
      writer.uint32(42).string(v!);
    }

    return writer;
  },

  decode(input: _m0.Reader | Uint8Array): ContextRule {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseContextRule();

    while (reader.pos < end) {
      const tag = reader.uint32();

      switch (tag >>> 3) {
        case 1:
          message.selector = reader.string();
          break;

        case 2:
          message.requested.push(reader.string());

        case 3:
          message.provided.push(reader.string());

        case 4:
          message.allowedRequestExtensions.push(reader.string());

        case 5:
          message.allowedResponseExtensions.push(reader.string());

        default:
          reader.skipType(tag & 7);
          break;
      }
    }

    return message;
  },

  fromJSON(object: any): ContextRule {
    return {
      selector: isSet(object.selector) ? String(object.selector) : "",
      requested: Array.isArray(object?.requested) ? object.requested.map((e: any) => String(e)) : [],
      provided: Array.isArray(object?.provided) ? object.provided.map((e: any) => String(e)) : [],
      allowedRequestExtensions: Array.isArray(object?.allowedRequestExtensions) ? object.allowedRequestExtensions.map((e: any) => String(e)) : [],
      allowedResponseExtensions: Array.isArray(object?.allowedResponseExtensions) ? object.allowedResponseExtensions.map((e: any) => String(e)) : []
    };
  },

  toJSON(message: ContextRule): unknown {
    const obj: any = {};
    message.selector !== undefined && (obj.selector = message.selector);

    if (message.requested) {
      obj.requested = message.requested.map(e => e);
    } else {
      obj.requested = [];
    }

    if (message.provided) {
      obj.provided = message.provided.map(e => e);
    } else {
      obj.provided = [];
    }

    if (message.allowedRequestExtensions) {
      obj.allowedRequestExtensions = message.allowedRequestExtensions.map(e => e);
    } else {
      obj.allowedRequestExtensions = [];
    }

    if (message.allowedResponseExtensions) {
      obj.allowedResponseExtensions = message.allowedResponseExtensions.map(e => e);
    } else {
      obj.allowedResponseExtensions = [];
    }

    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<ContextRule>, I>>(object: I): ContextRule {
    const message = createBaseContextRule();
    message.selector = object.selector ?? "";
    message.requested = object.requested?.map(e => e) || [];
    message.provided = object.provided?.map(e => e) || [];
    message.allowedRequestExtensions = object.allowedRequestExtensions?.map(e => e) || [];
    message.allowedResponseExtensions = object.allowedResponseExtensions?.map(e => e) || [];
    return message;
  }

};