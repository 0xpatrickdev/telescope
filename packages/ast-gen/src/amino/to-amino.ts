import * as t from '@babel/types';
import { BILLION, getFieldDimensionality, memberExpressionOrIdentifier, shorthandProperty } from '../utils';
import { Enum, Field, Interface, MessageSchema } from '../types';

export const toAminoDuration = (prop: string, scope: string[], aminoCasingFn: Function) => {
  const exp = t.binaryExpression(
    '*',
    memberExpressionOrIdentifier(scope),
    BILLION
  );
  exp.extra = { parenthesized: true };
  const value = t.callExpression(
    t.memberExpression(
      exp,
      t.identifier('toString')
    ),
    []
  )
  return t.objectProperty(t.identifier(aminoCasingFn(prop)), value);
};

export const toAminoCaseToCamel = (prop: string, scope: string[], aminoCasingFn: Function) => {
  if (prop === aminoCasingFn(prop) && scope.length === 1) {
    return shorthandProperty(prop);
  }
  return t.objectProperty(t.identifier(aminoCasingFn(prop)), memberExpressionOrIdentifier(scope))
};

export const toAminoInterface = (field: Field, ival: Interface, enums: Enum[], interfaces: Interface[], scope: string[], nested: number, aminoCasingFn: Function) => {
  const properties = ival.fields.map(field => {
    return toAminoParseField(field, enums, interfaces, scope, nested, aminoCasingFn)
  });
  return t.objectProperty(t.identifier(aminoCasingFn(field.name)),
    t.objectExpression(
      properties
    )
  );
};

export const toAminoInterfaceArray = (field: Field, ival: Interface, enums: Enum[], interfaces: Interface[], scope: string[], nested: number, aminoCasingFn: Function) => {
  const variable = 'el' + nested;
  const properties = ival.fields.map(field => {
    return toAminoParseField(field, enums, interfaces, [variable], nested + 1, aminoCasingFn)
  });

  const expr = t.callExpression(
    t.memberExpression(
      memberExpressionOrIdentifier(scope),
      t.identifier('map')
    ),
    [
      t.arrowFunctionExpression(
        [
          t.identifier(variable)
        ],
        t.objectExpression(
          properties
        )
      )
    ]
  );

  return t.objectProperty(t.identifier(aminoCasingFn(field.name)),
    expr
  );
}

export const toAminoLongToString = (prop: string, scope: string[], aminoCasingFn: Function) => {
  return t.objectProperty(t.identifier(aminoCasingFn(prop)),
    t.callExpression(
      t.memberExpression(memberExpressionOrIdentifier(scope), t.identifier('toString')),
      [])
  )
};

export const toAminoCoin = (prop: string, scope: string[], aminoCasingFn: Function) => {
  const value = t.objectExpression([
    t.objectProperty(t.identifier('denom'), t.memberExpression(
      memberExpressionOrIdentifier(scope),
      t.identifier('denom'),
    )),
    t.objectProperty(
      t.identifier('amount'),
      t.callExpression(
        t.memberExpression(
          t.callExpression(
            t.memberExpression(
              t.identifier('Long'),
              t.identifier('fromNumber')
            ),
            [
              t.memberExpression(
                memberExpressionOrIdentifier(scope),
                t.identifier('amount')
              )
            ]
          ),
          t.identifier('toString')
        ),
        []
      )
    )
  ]);
  return t.objectProperty(t.identifier(aminoCasingFn(prop)), value);
};

export const toAminoHeight = (prop: string, scope: string[], aminoCasingFn: Function) => {
  const value = t.objectExpression(
    [
      t.objectProperty(
        t.identifier(aminoCasingFn('revision_height')),
        t.optionalCallExpression(
          t.optionalMemberExpression(
            t.callExpression(
              t.identifier('omitDefault'),
              [
                t.memberExpression(
                  memberExpressionOrIdentifier(scope),
                  t.identifier('revisionHeight')
                )
              ]
            ),
            t.identifier('toString'),
            false,
            true
          ),
          [],
          false
        )
      ),
      //
      t.objectProperty(
        t.identifier(aminoCasingFn('revision_number')),
        t.optionalCallExpression(
          t.optionalMemberExpression(
            t.callExpression(
              t.identifier('omitDefault'),
              [
                t.memberExpression(
                  memberExpressionOrIdentifier(scope),
                  t.identifier('revisionNumber')
                )
              ]
            ),
            t.identifier('toString'),
            false,
            true
          ),
          [],
          false
        )
      )
    ]
  );

  const cond = t.conditionalExpression(
    memberExpressionOrIdentifier(scope),
    value,
    t.objectExpression([])
  );

  return t.objectProperty(t.identifier(aminoCasingFn(prop)), cond);
};

export const toAminoParseField = (field: Field, enums: Enum[], interfaces: Interface[], scope = [], nested = 0, aminoCasingFn: Function) => {

  const newScope = [field.name, ...scope];
  const { typeName, dimensions, isArray } = getFieldDimensionality(field);

  // special cases
  switch (field.type) {
    case 'Coin':
      return toAminoCoin(field.name, newScope, aminoCasingFn);
    case 'Duration':
      return toAminoDuration(field.name, newScope, aminoCasingFn);
    case 'Height':
      return toAminoHeight(field.name, newScope, aminoCasingFn);
    default:
  }

  // TODO QUESTION do we need to handle enums?

  // definitions/interfaces
  const ival: Interface = interfaces.find(e => e.name === typeName);
  if (ival) {
    if (!isArray) {
      return toAminoInterface(field, ival, enums, interfaces, newScope, nested, aminoCasingFn);
    } else if (dimensions === 0) {
      return toAminoInterfaceArray(field, ival, enums, interfaces, newScope, nested, aminoCasingFn);
    }
  }

  switch (field.type) {
    case 'Long':
      return toAminoLongToString(field.name, newScope, aminoCasingFn);
    default:
      return toAminoCaseToCamel(field.name, newScope, aminoCasingFn);
  }
};

export const toAmino = (schema: MessageSchema, enums: Enum[], interfaces: Interface[], aminoCasingFn: Function) => {

  const TypeName = schema.name;

  const toAminoParams = t.objectPattern(
    schema.fields.map((field) => t.objectProperty(t.identifier(field.name), t.identifier(field.name), false, true))
  );
  toAminoParams.typeAnnotation = t.tsTypeAnnotation(t.tsTypeReference(t.identifier(TypeName)))

  const toAmino = t.arrowFunctionExpression(
    [
      toAminoParams
    ],
    t.blockStatement([
      t.returnStatement(
        t.objectExpression(
          schema.fields.map((field) => toAminoParseField(field, enums, interfaces, undefined, undefined, aminoCasingFn))
        )
      )
    ])
  );

  toAmino.returnType = t.tsTypeAnnotation(t.tsIndexedAccessType(
    t.tsTypeReference(t.identifier('Amino' + TypeName)),
    t.tsLiteralType(t.stringLiteral('value'))
  ));

  return toAmino;
}
