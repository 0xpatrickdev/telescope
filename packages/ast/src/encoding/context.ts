import { ProtoStore, ProtoRef } from '@osmonauts/proto-parser';

export interface ParseContext {
    utils: Record<string, boolean>;
    addUtil: Function;
}

export class GenericParseContext implements ParseContext {
    imports: Record<string, object> = {};
    utils: Record<string, boolean> = {};

    addUtil(util) {
        this.utils[util] = true;
    }
}
export class AminoParseContext implements ParseContext {
    store: ProtoStore;
    ref: ProtoRef;

    utils: Record<string, boolean> = {};

    constructor(
        ref: ProtoRef,
        store: ProtoStore
    ) {
        this.ref = ref;
        this.store = store;
    }

    addUtil(util) {
        this.utils[util] = true;
    }
}

export class ProtoParseContext implements ParseContext {
    imports: Record<string, object> = {};
    utils: Record<string, boolean> = {};

    addUtil(util) {
        this.utils[util] = true;
    }
}

