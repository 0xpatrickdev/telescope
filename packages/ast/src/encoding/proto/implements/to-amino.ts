import * as t from '@babel/types';
import { arrowFunctionExpression, identifier } from '../../../utils';
import { ProtoParseContext } from "../../context";

export const createInterfaceToAminoHelper = (
    context: ProtoParseContext,
    functionName: string
) => {

    const switchCases = [
        t.switchCase(
            t.stringLiteral('/cosmos.gov.v1beta1.TextProposal'),
            [
                t.blockStatement([
                    // decoded data
                    t.variableDeclaration('const', [
                        t.variableDeclarator(
                            t.identifier('data'),
                            t.callExpression(
                                t.memberExpression(
                                    t.identifier('TextProposal'),
                                    t.identifier('decode')
                                ),
                                [
                                    t.memberExpression(
                                        t.identifier('content'),
                                        t.identifier('value')
                                    )
                                ]
                            )
                        )
                    ]),
                    // return
                    t.returnStatement(
                        t.objectExpression([
                            t.objectProperty(
                                t.identifier('type'),
                                t.stringLiteral('cosmos-sdk/TextProposal')
                            ),
                            t.objectProperty(
                                t.identifier('value'),
                                t.callExpression(
                                    t.memberExpression(
                                        t.identifier('TextProposal'),
                                        t.identifier('toAmino')
                                    ),
                                    [
                                        t.identifier('data')
                                    ]
                                )
                            )
                        ])
                    )
                ])
            ])
    ];

    return t.exportNamedDeclaration(
        t.variableDeclaration(
            'const',
            [
                t.variableDeclarator(
                    t.identifier(functionName),
                    t.arrowFunctionExpression(
                        [
                            identifier(
                                'content',
                                t.tsTypeAnnotation(
                                    t.tsTypeReference(
                                        t.identifier('Any')
                                    )
                                )
                            )
                        ],
                        t.blockStatement([
                            t.switchStatement(
                                t.memberExpression(
                                    t.identifier('content'),
                                    t.identifier('typeUrl')
                                ),
                                [
                                    ...switchCases,
                                    // default
                                    t.switchCase(
                                        null,
                                        [
                                            t.returnStatement(
                                                t.callExpression(
                                                    t.memberExpression(
                                                        t.identifier('Any'),
                                                        t.identifier('toAmino')
                                                    ),
                                                    [
                                                        t.identifier('content')
                                                    ]
                                                )
                                            )
                                        ]
                                    )
                                ]
                            )
                        ])
                    )

                )
            ]
        )
    )
};