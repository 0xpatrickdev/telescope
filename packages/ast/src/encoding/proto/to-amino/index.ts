import * as t from '@babel/types';
import { getFieldOptionality, getOneOfs } from '..';
import { identifier, objectMethod } from '../../../utils';
import { ProtoParseContext } from '../../context';
import { ProtoField, ProtoType } from '@osmonauts/types';
import { arrayTypes, toAminoJSON } from './utils';
import { pascal } from 'case';

const needsImplementation = (name: string, field: ProtoField) => {
    throw new Error(`need to implement toAminoJSON (${field.type} rules[${field.rule}] name[${name}])`);
}

export interface ToAminoJSONMethod {
    context: ProtoParseContext;
    field: ProtoField;
    isOptional: boolean;
}

export const toAminoJSONMethodFields = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const oneOfs = getOneOfs(proto);
    const fields = Object.keys(proto.fields ?? {}).reduce((m, fieldName) => {
        const field = {
            name: fieldName,
            ...proto.fields[fieldName]
        };

        const isOneOf = oneOfs.includes(fieldName);
        const isOptional = getFieldOptionality(context, field, isOneOf);

        const args: ToAminoJSONMethod = {
            context,
            field,
            isOptional
        };

        if (field.rule === 'repeated') {
            switch (field.type) {
                case 'string':
                    return [...m, toAminoJSON.array(args, arrayTypes.string())];
                case 'bytes':
                    return [...m, toAminoJSON.array(args, arrayTypes.bytes(args))];
                case 'bool':
                    return [...m, toAminoJSON.array(args, arrayTypes.bool())];
                case 'double':
                    return [...m, toAminoJSON.array(args, arrayTypes.double())];
                case 'float':
                    return [...m, toAminoJSON.array(args, arrayTypes.float())];
                case 'int32':
                    return [...m, toAminoJSON.array(args, arrayTypes.int32())];
                case 'sint32':
                    return [...m, toAminoJSON.array(args, arrayTypes.sint32())];
                case 'uint32':
                    return [...m, toAminoJSON.array(args, arrayTypes.uint32())];
                case 'fixed32':
                    return [...m, toAminoJSON.array(args, arrayTypes.fixed32())];
                case 'sfixed32':
                    return [...m, toAminoJSON.array(args, arrayTypes.sfixed32())];
                case 'int64':
                    return [...m, toAminoJSON.array(args, arrayTypes.int64(args))];
                case 'sint64':
                    return [...m, toAminoJSON.array(args, arrayTypes.sint64(args))];
                case 'uint64':
                    return [...m, toAminoJSON.array(args, arrayTypes.uint64(args))];
                case 'fixed64':
                    return [...m, toAminoJSON.array(args, arrayTypes.fixed64(args))];
                case 'sfixed64':
                    return [...m, toAminoJSON.array(args, arrayTypes.sfixed64(args))];
                default:
                    switch (field.parsedType.type) {
                        case 'Enum':
                            return [...m, toAminoJSON.array(args, arrayTypes.enum(args))];
                        case 'Type':
                            return [...m, toAminoJSON.array(args, arrayTypes.type(args))];
                    }
                    return needsImplementation(fieldName, field);
            }

        }

        if (field.keyType) {
            switch (field.keyType) {
                case 'string':
                case 'int32':
                case 'sint32':
                case 'uint32':
                case 'fixed32':
                case 'sfixed32':
                case 'int64':
                case 'sint64':
                case 'uint64':
                case 'fixed64':
                case 'sfixed64':
                    return [...m, ...toAminoJSON.keyHash(args)];
                default:
                    return needsImplementation(fieldName, field);
            }
        }

        switch (field.type) {
            case 'string':
                return [...m, toAminoJSON.string(args)];
            case 'double':
                return [...m, toAminoJSON.double(args)];
            case 'float':
                return [...m, toAminoJSON.float(args)];
            case 'bytes':
                return [...m, toAminoJSON.bytes(args)];
            case 'bool':
                return [...m, toAminoJSON.bool(args)];
            case 'int32':
                return [...m, toAminoJSON.int32(args)];
            case 'sint32':
                return [...m, toAminoJSON.sint32(args)];
            case 'uint32':
                return [...m, toAminoJSON.uint32(args)];
            case 'fixed32':
                return [...m, toAminoJSON.fixed32(args)];
            case 'sfixed32':
                return [...m, toAminoJSON.sfixed32(args)];
            case 'int64':
                return [...m, toAminoJSON.int64(args)];
            case 'sint64':
                return [...m, toAminoJSON.sint64(args)];
            case 'uint64':
                return [...m, toAminoJSON.uint64(args)];
            case 'fixed64':
                return [...m, toAminoJSON.fixed64(args)];
            case 'sfixed64':
                return [...m, toAminoJSON.sfixed64(args)];
            case 'google.protobuf.Duration':
            case 'Duration':
                return [...m, toAminoJSON.duration(args)];
            case 'google.protobuf.Timestamp':
            case 'Timestamp':
                return [...m, toAminoJSON.timestamp(args)];
            default:
                switch (field.parsedType.type) {
                    case 'Enum':
                        return [...m, toAminoJSON.enum(args)];
                    case 'Type':
                        return [...m, toAminoJSON.type(args)];
                }
                return needsImplementation(fieldName, field);
        }
    }, []);
    return fields;
};

export const toAminoJSONMethod = (context: ProtoParseContext, name: string, proto: ProtoType) => {
    const fields = toAminoJSONMethodFields(context, name, proto);
    let varName = 'message';
    if (!fields.length) {
        varName = '_';
    }

    const SDKTypeName =
        [name, 'SDKType']
            .filter(Boolean).join('')


    return objectMethod('method',
        t.identifier('toAmino'),
        [
            identifier(
                varName,
                t.tsTypeAnnotation(
                    t.tsTypeReference(
                        t.identifier(name)
                    )
                )
            )
        ],
        t.blockStatement([
            // t.variableDeclaration('const', [
            //     t.variableDeclarator(
            //         t.identifier('hoistedAsts'),
            //         t.nullLiteral()
            //     )
            // ]),

            t.variableDeclaration(
                'const',
                [
                    t.variableDeclarator(
                        identifier('obj', t.tsTypeAnnotation(t.tsAnyKeyword())),
                        t.objectExpression([])
                    )
                ]
            ),

            ...fields,

            // RETURN 
            t.returnStatement(t.identifier('obj'))

        ]),
        false,
        false,
        false,
        t.tsTypeAnnotation(
            t.tsTypeReference(
                t.identifier(SDKTypeName)
            )
        )
    );
};
