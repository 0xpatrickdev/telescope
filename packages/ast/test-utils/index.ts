import generate from '@babel/generator';
import { ProtoStore, getNestedProto, parseProto } from '@osmonauts/proto-parser'
import { ProtoType } from '../src/encoding/proto/types';
import { ParseContext } from '../src/encoding/context';

export const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}
export const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

export const prepareContext = (store: ProtoStore, protoFile: string) => {
    const ref = store.findProto(protoFile);
    const traversed = ref.traversed
    if (!traversed) {
        throw new Error('not traversed!');
    }
    const proto: Record<string, ProtoType> = getNestedProto(traversed);
    const protos: ProtoType[] = Object.values(proto).filter(
        proto => proto.name.startsWith('Msg')
            &&
            !proto.name.endsWith('Response')
            &&
            proto.name !== 'Msg'
    );

    const context = new ParseContext(ref, store);

    return {
        context,
        root: traversed,
        protos
    };
};
