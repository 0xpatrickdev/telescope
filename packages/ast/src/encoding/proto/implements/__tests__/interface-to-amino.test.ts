import { expectCode, printCode, getTestProtoStore } from '../../../../../test-utils/'
import { ProtoParseContext } from '../../../context';
import { createInterfaceToAminoHelper } from '../to-amino';

const store = getTestProtoStore();
store.options.prototypes.implementsAcceptsAny = true;
store.options.aminoEncoding.enabled = true;
store.options.aminoEncoding.useRecursiveV2encoding = true;
store.traverseAll();

describe('ProposalContent', () => {
    const txRef = store.findProto('cosmos/gov/v1/tx.proto');
    const txContext = new ProtoParseContext(txRef, store, store.options);
    it('ProposalContent', () => {
        expectCode(createInterfaceToAminoHelper(
            txContext,
            'ProposalContentI_toAmino'
        ));
    });
});