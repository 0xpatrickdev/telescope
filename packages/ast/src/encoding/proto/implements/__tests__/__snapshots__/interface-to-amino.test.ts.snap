// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProposalContent ProposalContent 1`] = `
"export const ProposalContentI_toAmino = (content: Any) => {
  switch (content.typeUrl) {
    case "/cosmos.gov.v1beta1.TextProposal":
      {
        const data = TextProposal.decode(content.value);
        return {
          type: "cosmos-sdk/TextProposal",
          value: TextProposal.toAmino(data)
        };
      }

    default:
      return Any.toAmino(content);
  }
};"
`;
